### ISSUE TITLE
Web Assembly port

### ISSUE THREAD

**Issue by pcwalton** [servo/pathfinder#155]
It would be nice to have a Web Assembly port, like Google's CanvasKit for Skia.

**Comment by asajeffrey** [comment]
Yes please!

**Comment by asajeffrey** [comment]
There might be a bit of work getting the multithreaded story sorted out, it would be nice if we could use post message rather than SharedArrayBuffer for communication.

**Comment by pcwalton** [comment]
Web Assembly is one of the reasons behind the Executor and RenderCommandListener abstractions. (The other one is WebRender, which is hesitant to add any more threads, at least in Gecko.) Core Pathfinder no longer spawns any threads. For a single-threaded workload you would buffer RenderCommands into a vector, then send them to a Renderer all at once instead of using a SceneProxy. This is likely what a Web Assembly port will want to do.

**Comment by est31** [comment]
I'd love to help with this. I need some project to find out how ready the wasm ecosystem is and this seems like it could be that project. @pcwalton do you have a list of items to work on?

**Comment by pcwalton** [comment]
@est31 Here are the remaining blockers:

- [ ] HarfBuzz needs to compile with wasm
- [ ] FreeType needs to compile with wasm
- [ ] Check skribo to make sure it compiles
- [ ] Check Pathfinder to make sure it compiles
- [ ] The `gl` crate needs a WebGL implementation
- [ ] Canvas C API needs to be finished
- [ ] JavaScript bindings need to be written

**Comment by est31** [comment]
@pcwalton thanks for the list.

I'm not sure I want to do the HarfBuzz/Freetype parts because I don't want to attempt installing emscripten again. It's very tedious to install and set up. The js embedding for wasi right now seems to build on emscripten itself :/. So maybe I should work on #164 first?

>  Canvas C API needs to be finished

Instead of the C API, I'd have used wasm-bindgen instead. It allows you to expose rust-y APIs to javascript.

**Comment by pcwalton** [comment]
Any and all help is appreciated! I'll probably get to FreeType and HB if someone else doesn't do it first.

As for the C API, we need it for other reasons, for example @toolness' Unity plugin, so I figured we might as well use it for the wasm port as well. But I don't feel strongly about it either way.

**Comment by est31** [comment]
Hmm yeah I just realized that #164 is larger than I thought. It requires changes in font-kit as well as skribo or requires replacing skribo use entirely both of which is probably involved and it's probably easier for you given you know both much better than me.

I'd love to help though after #164 is resolved! Of course, only if you haven't done WASM support by then already :).

**Comment by jrowny** [comment]
FreeType and Harfbuzz compile without any modification via emscripten. We'd have to compile the whole thing via `wasm32-unknown-emscripten` then right? I'm a bit new to Rust, but I'd love to help with this as well if I can.

**Comment by jrowny** [comment]
Also, won't SIMD be a problem?

**Comment by est31** [comment]
@jrowny feel free to take this issue :)

**Comment by jnetterf** [comment]
I have a PR which adds a [glow](https://github.com/grovesNL/glow) implementation of the device abstraction: https://github.com/servo/pathfinder/pull/243

This allows parts of pathfinder to target WebGL 2 on `wasm32-unknown-unknown`. I'd appreciate feedback.

If this works, I think this leaves getting FreeType and Harfbuzz to compile without emscripten as the main challenge.

**Comment by s3bk** [comment]
I would like to report success!
[DEMO](https://grafeia.github.io/grafeia-wasm/)

Modified pathfinder repo: https://github.com/s3bk/pathfinder
Grafeia repo: https://github.com/grafeia/grafeia

**Comment by pcwalton** [comment]
Wow!

**Comment by boulabiar** [comment]
Excellent !
Does it work with complex languages like Arabic?

**Comment by s3bk** [comment]
@boulabiar I plan to support right to left scripts eventually. The pdf viewer using the font crate renders Arabic text correct.
What I may not implement for a while is embeddings, and instead require a new paragraph when the direction is switched.

**Comment by s3bk** [comment]
Upon request from #rust-offtopic, here is the PDF viewer running in wasm:

https://pdf-rs.github.io/view-wasm/

(use ◁ and ▷ keys to navigate)

**Comment by CoreyDotCom** [comment]
Nice job @s3bk! Any chance you can hoist a sample up so that it's straight forward to infer performance of the tiger sample? e.g. Hard to tell if it's redrawing in full as you pan or if things are cached at that point. Looking to check out what the perf/cost is for dynamically changing path data / animation in the wasm port.

**Comment by s3bk** [comment]
@CoreyDotCom the scene of the tiger is cached, but everything is redrawn every frame (that something changed).

**Comment by s3bk** [comment]
@CoreyDotCom https://github.com/s3bk/webgl_demo/
https://s3bk.github.io/webgl_demo/

**Comment by leoyanggit** [comment]
> https://github.com/s3bk/pathfinder



> I would like to report success!
> [DEMO](https://grafeia.github.io/grafeia-wasm/)
> 
> Modified pathfinder repo: https://github.com/s3bk/pathfinder
> Grafeia repo: https://github.com/grafeia/grafeia
> 
> I am using [vector](https://github.com/s3bk/vector) to draw things and [font](https://github.com/pdf-rs/font) to parse fonts and draw text.

@s3bk WoW looks promising. I wanted to try it out. Is there instruction about how to build pathfinder with your change to wasm?

**Comment by s3bk** [comment]
@leoyanggit 
look at the Makefile in https://github.com/s3bk/webgl_demo/
You can add a [patch] section to use `servo/pathfinder`

**Comment by leoyanggit** [comment]
@s3bk Thanks! I successfully compiled and ran webgl_demo. I'm currently looking into text rendering. You mentioned you are using https://github.com/pdf-rs/font to render text. I'm wondering you have a demo that uses it along with pathfinder.

**Comment by s3bk** [comment]
@leoyanggit you are lucky. I just implemented some for pathfinder_canvas
https://github.com/s3bk/pathfinder/commit/5b003f456ff01e0747a020e565dc3aa88771ce46
```
pathfinder_canvas = { git = "https://github.com/s3bk/pathfinder", features = ["pf-simple-text"] }
```
Demo using the canvas api from python:

https://github.com/s3bk/pathfinder_python

**Comment by s3bk** [comment]
woops. WebAssembly … the canvas text stuff should still. work.

**Comment by leoyanggit** [comment]
If I use pf-text feature `pathfinder_canvas = { git = "https://github.com/servo/pathfinder/", features = ["pf-text"] }` in my project, `wasm-pack build -t no-modules --release` will try to compile `servo-freetype-sys` with wasm32-unknown-unknown target. But it failed with the following error. It looks like wasm-pack calling my clang directly with wasm32-unknown-unknown target. I'm new to Rust and WebAssembly. I'm wondering if I should expect wasm-pack call emscripten to compile `servo-freetype-sys` for this case.

`
    Run Build Command(s):/usr/bin/make cmTC_118b1/fast && /Library/Developer/CommandLineTools/usr/bin/make  -f CMakeFiles/cmTC_118b1.dir/build.make CMakeFiles/cmTC_118b1.dir/build
    Building C object CMakeFiles/cmTC_118b1.dir/testCCompiler.c.o
    /usr/bin/clang   -ffunction-sections -fdata-sections -fPIC --target=wasm32-unknown-unknown  -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk   -o CMakeFiles/cmTC_118b1.dir/testCCompiler.c.o   -c /Users/lyang/test/target/wasm32-unknown-unknown/release/build/servo-freetype-sys-8979fe94f5a2c82d/out/build/CMakeFiles/CMakeTmp/testCCompiler.c
    error: unable to create target: 'No available targets are compatible with triple "wasm32-unknown-unknown"'
    1 error generated.
    make[1]: *** [CMakeFiles/cmTC_118b1.dir/testCCompiler.c.o] Error 1
    make: *** [cmTC_118b1/fast] Error 2`

**Comment by s3bk** [comment]
`pf-text` is the old freetype and harfbuzz based impl.
`pf-simple-text` is new and uses my font parser.

`pf-text` will also override `pf-simple-text` so if anything enables that, you get the above problem

**Comment by s3bk** [comment]
If posslble, could we move this to #pathfinder:matrix.org?
I expect a few more problems… and this issue isn't really a good place.

**Comment by pcwalton** [comment]
We have a wasm port now. Let's close this in favor of more specific items.

**Comment by murtyjones** [comment]
> @leoyanggit you are lucky. I just implemented some for pathfinder_canvas [s3bk@5b003f4](https://github.com/s3bk/pathfinder/commit/5b003f456ff01e0747a020e565dc3aa88771ce46)
> 
> ```
> pathfinder_canvas = { git = "https://github.com/s3bk/pathfinder", features = ["pf-simple-text"] }
> ```
> 
> Demo using the canvas api from python:
> 
> https://github.com/s3bk/pathfinder_python

This is very helpful. Are there any plans to port `pf-simple-text` back to `pathfinder`?

**Comment by s3bk** [comment]
@murtyjones: To be honest, unlikely. I am not using it. If you really need the canvas api, making a fork to use svg_text will be the easiest.
